### ISSUE TITLE
CPP guards for very old Cabal versions in the tests: what do we do?

### ISSUE THREAD

**Issue by Kleidukos** [haskell/cabal#8998]
I found some quite old Cabal versions being taken into account in the tests, especially here: https://github.com/haskell/cabal/blob/c3f92c4c39ad1ed643cf7837182d057f59f3396e/Cabal-tests/tests/custom-setup/CabalDoctestSetup.hs#L124

Do we still have an obligation to support Cabal 1.25 in our test suite?

**Comment by ffaf1** [comment]
Tag `Cabal-v1.24.2.0` commit says `Thu Dec 8 02:43:11 2016 +0000`, so 6½ years ago. I can imagine relatively old Haskell being deployed somewhere, I can’t imagine a situation where they would want to build cabal head.

**Comment by andreabedini** [comment]
I am planning to drop support for anything earlier than what is promised (of the top of my head: 5-year old GHCs for Cabal and 3-year old GHCs for a cabal-install)

There's lot of ifdefs, special behaviours and work arounds for GHC<8!

I have a WIP in a branch, but I need to double check as I might have got my GHC versions wrong and  removed support for one version later 😅 I guess we can wait until September to merge 🤣

**Comment by Mikolaj** [comment]
Yes, that's a great idea. Let's clean up the cruft. But let's remember new cabal needs to *work* with even older GHCs, that are not supposed to be able to compile cabal. So if any of the ifdefs relate to cabal API or cabal-GHC protocol changes, that requires an extra discussion.

**Comment by andreabedini** [comment]
> But let's remember new cabal needs to _work_ with even older GHCs, that are not supposed to be able to compile cabal.

Oh, thank you @Mikolaj for making this distinction! I have the impression it is not properly captured in CONTRIBUTING.md, which says, generically, _support_ and seems to identify it with _buildable out-of-the-box_.

> Our GHC support window is five years for the Cabal library and three years for cabal-install: that is, the Cabal library must be buildable out-of-the-box with the dependencies that shipped with GHC for at least five years.

As you allude to, what "support" entitles can be tricky. E.g. GHC 8.0.2 (2017-01-11) shipped with Cabal-1.24.2.0; but one can obtain cabal-install-3.10.1.0 (e.g. compiled with GHC 9.6.2) which can use its internal Cabal-3.10.1.0 to "drive" GHC 8.0.2. In another scenario (e.g. a custom setup) cabal-install-3.10.1.0 will have to talk to Cabal-1.24.2.0 (I think?).

As far as I can see , "support" can have at least these meanings:

- Cabal _can be built with_ GHC X.Y.Z _boot packages_
- cabal-install _can be built with_ GHC X.Y.Z
- Cabal _can drive_ with GHC X.Y.Z
- cabal-install _can drive_ Cabal A.B.C.D

I think we need to clarify this topic so we can be all on the same page.

**Comment by fgaz** [comment]
> Cabal can drive with GHC X.Y.Z

Currently we support ghc >= 7.0.1 && < 9.8:

https://github.com/haskell/cabal/blob/a191f0a2bf72894b3d830ca3ee4ea111d96b6f65/Cabal/src/Distribution/Simple/GHC.hs#L168-L182

> cabal-install can drive Cabal A.B.C.D

Currently Cabal >= 1.20

https://github.com/haskell/cabal/blob/a191f0a2bf72894b3d830ca3ee4ea111d96b6f65/cabal-install/src/Distribution/Client/ProjectPlanning.hs#L1390-L1438

**Comment by philderbeast** [comment]
> Currently we support ghc >= 7.0.1 && < 9.8:

The [announcement for ghc-7.0.1](https://mail.haskell.org/pipermail/haskell/2010-November/022376.html) was made 2010-11-15, that is:

> 12 years, 7 months, 3 days ago
> SOURCE: [Days Calculator: Days Between Two Dates](https://www.timeanddate.com/date/durationresult.html?m1=11&d1=15&y1=2010&m2=6&d2=18&y2=2023)

**Comment by ulysses4ever** [comment]
@philderbeast 

> The [announcement for ghc-7.0.1](https://mail.haskell.org/pipermail/haskell/2010-November/022376.html) was made 2010-11-15, that is: 12 years, 7 months, 3 days ago

Do you mean we should drop it on grounds of being too old? But we're not doing anything special, as this is the version of GHC we can "drive" rather than compile with (see discussion above). So, I don't think we have to do anything about that. Maybe apart from checking that this is still the case for newer `Cabal`s. I think this is what ["Validate old ghcs"](https://github.com/haskell/cabal/blob/master/.github/workflows/validate.yml#L175) job does?..

---

@ All,

should we put compatibility data discussed in this thread in a visible place? E.g. README.md or CONTRIBUTING.md or anything else?

**Comment by Kleidukos** [comment]
Yes I would rather appreciate if there was a clear and unique source of information regarding this.

**Comment by philderbeast** [comment]
Getting rid of some very old `CPP` `#if` conditionals could help us when we follow up on #9041, likely with these two:

```yaml
# Warnings currently triggered by your code
- ignore: {name: "Unused LANGUAGE pragma"} # 121 hints
- ignore: {name: "Use fewer imports"} # 13 hints
```

**Comment by ulysses4ever** [comment]
@philderbeast that would be very good, thank you for clarifying! But support for driving GHC 7.0 shouldn't cost us any CPP as far as I understand.

**Comment by philderbeast** [comment]
Thanks @ulysses4ever and others for mentioning GHC versions that cabal can drive. Seems like we support a hell of a lot of GHC versions for building and testing of cabal the project itself. As a user of cabal, I'll get it via ghcup. For developing and maintaining cabal, do we need to support building with versions older than `ghc-9.4.8`?

https://github.com/haskell/cabal/blob/ce5d0f7a1a8481e2ed599f81c8a8387dccb3a979/.github/workflows/validate.yml#L52

**Comment by Mikolaj** [comment]
I think there's much more than exe users and cabal devs. There are distribution packagers, devs of tools using Cabal, industry power users that compile patched cabal-install using their ancient GHCs (I may be making that one up, but who knows), etc. Unless it's a considerable burden, I'd keep the current backward compatibility windows whatever they are (2 and 5 years? this must be written somewhere).

**Comment by philderbeast** [comment]
I see that `ghc-8.0.1` was released "21 May 2016" from [GHC status](https://gitlab.haskell.org/ghc/ghc/-/wikis/GHC-status). Can we bump the `CPP` guards forward to this point in time?

```
#if MIN_VERSION_base(4,9,0)
```

This would allow us to do things like switch from `-fwarn-*` and `-fno-warn-*` to `-W*` and `-Wno-*`.

> <p>In GHC &lt; 8 the syntax for <code class="docutils literal notranslate"><span class="pre">-W⟨wflag⟩</span></code> was <code class="docutils literal notranslate"><span class="pre">-f<span class="highlighted">warn</span>-⟨wflag⟩</span></code>
> (e.g. <code class="docutils literal notranslate"><span class="pre">-f<span class="highlighted">warn</span>-incomplete-patterns</span></code>).
> This spelling is deprecated, but still accepted for backwards compatibility.
> Likewise, <code class="docutils literal notranslate"><span class="pre">-Wno-⟨wflag⟩</span></code> used to be <code class="docutils literal notranslate"><span class="pre"><span class="highlighted">fno</span>-<span class="highlighted">warn</span>-⟨wflag⟩</span></code>
> (e.g. <code class="docutils literal notranslate"><span class="pre">-<span class="highlighted">fno</span>-<span class="highlighted">warn</span>-incomplete-patterns</span></code>).</p>
>
> SOURCE: [GHC Users Guide - Warnings and sanity-checking](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/using-warnings.html#warnings-and-sanity-checking).

**Comment by ulysses4ever** [comment]
Indeed, GHC 8.0 and 8.2 are out of our support window.

**Comment by Mikolaj** [comment]
Sure, let's do it, but let's remember to mention it prominently in the changelog file for the PR to make users aware.

**Comment by philderbeast** [comment]
Here are the current CPP macros we use for base versions:

```
$ grep -R -oP -h 'MIN_VERSION_base\(.*\)' */**/*.hs | sort | uniq
MIN_VERSION_base(4,0,0)
MIN_VERSION_base(4,11,0)
MIN_VERSION_base(4,18,0)
MIN_VERSION_base(4,20,0)
MIN_VERSION_base(4,7,0)
```

**Comment by philderbeast** [comment]
Most of the base CPP conditions, "base 4.13 and older", were removed in #10092.

**Comment by philderbeast** [comment]
We're still validating older versions of GHC. Do we still need to do that?

```
$ grep -R -E '"8\.(0|2)' .github/workflows/*.yml
.github/workflows/validate.yml:          ["8.4.4", "8.2.2", "8.0.2"]
```

https://github.com/haskell/cabal/blob/2040d210563e55f7bacd2bd4b7572ccbf25f6c8c/.github/workflows/validate.yml#L266-L277

**Comment by Mikolaj** [comment]
> We're still validating older versions of GHC. Do we still need to do that?

As a rule, if there's no good reasons to drop older GHCs. we don't.
